include/parlay/internal/uninitialized_iterator.h

```diff
@@ -0,0 +1,158 @@
+
+#ifndef PARLAY_INTERNAL_UNINITIALIZED_ITERATOR_H_
+#define PARLAY_INTERNAL_UNINITIALIZED_ITERATOR_H_
+
+#include <iterator>
+#include <type_traits>
+
+#include "../range.h"
+#include "../type_traits.h"
+
+namespace parlay {
+namespace internal {
+
+// Given a container of uninitialized<T>, you can wrap its iterators with
+// uninitialized_iterator_adapter to get an iterator whose value type is T!
+//
+// The resulting iterator will have the same iterator category as Iterator.
+template<typename Iterator>
+class uninitialized_iterator_adapter {
```

Comment:
nit: ISO (and me) spell it `adaptor`, although the dictionary accepts both ways.
https://en.cppreference.com/w/cpp/experimental/resource_adaptor